### ISSUE TITLE
Incorporate microfounded demand into supply estimation

### ISSUE THREAD

**Issue by tgerarden** [tgerarden/solar-tariffs-microfounded-demand#1]
**Goal:** Aggregate existing estimated demand elasticities from the downstream solar markets (utility and residential/commercial), insert it into the firm FOC to recover estimates of firms' costs, and then check the results. (this is just the first step of many)

**Note:** @hapham-econ i set up this repo inside our existing dropbox. it's not a perfect solution but it seemed like a compromise between the extremes of starting a totally separate git repo and not using git. my hope is that by putting this code in a subfolder, you will be the primary person touching it, and we won't end up with conflicts or inadvertent changes committed to the repo due to dropbox syncing.

the code this is in: `~/Dropbox/SolarTariff/regressions/cournot/microfounded_supply_microfounded_demand`

(please leave the `aggregate_demand` folder and content as is!!)

### To do:
- modify the code in section 7) of `estimate_model.R` to:
  - [ ] read in the utility-scale elasticities from `SolarTariff/regressions/utility_demand_estimation/utility_demand_elasticities.csv`. for now let's use the values in `util_elasticity_other_iv`. please do this by creating a macro/object at the beginning of the `estimate_model.R` script that stores "util_elasticity_other_iv" and use that to select the relevant column in the details of the code since we might want to change this specification choice later. (see code for examples)
  - [ ] read in the residential/commercial-scale elasticities from `SolarTariff/counterfactual/elasticity.xlsx`. there are no dates in that file but the dates correspond to the same quarterly dates as the utility data (they come from a coauthor's code). you can just rbind them for now but you'll have to drop the final element of the vector since the other data only go through 2020 Q3, not 2020 Q4
  - [ ] join these two sets of electricities with the object `demand_share_data` in the environment (created at the beginning when the script `prep_cournot_data.R` is sourced) to compute the weighted average elasticity. you'll need to combine `share_residential` and `share_commercial`. then use the shares to compute the weighted average elasticity for each time period. assign that to alpha and join it to `merged_data`
  - [ ] then use the first bit of code in 8) to compute marginal costs (`gross_mc`) and report back with basic summary statistics on the marginal costs as well as a plot of price and marginal cost (color by firm) over time

please also follow the convention in the code of using relative paths not absolute paths when reading in files, etc.

let me know if you have any questions

**Comment by tgerarden** [comment]
@hapham-econ this first step will be trivial and i probably could have done it in the time it took to type this up. but as long as the results are workable you can build on this to dive into more interesting and time-intensive work. thanks!

**Comment by hapham-econ** [comment]
@tgerarden I will work on this tomorrow morning. Is that okay?

**Comment by hapham-econ** [comment]
@tgerarden Update: I ran into a technical issue. I could not connect to your computer using the VNC app. I am connected to the Cornell VPN but it still doesn't work. I will try to connect from my office first thing in the morning.

![image](https://github.com/tgerarden/solar-tariffs-microfounded-demand/assets/90290294/4c4416e9-9207-4c1b-b662-80afacf7d7ec)

**Comment by tgerarden** [comment]
@hapham-econ it looks like this is an issue with the computer, not your connection. the address changed so i needed to edit the address to connect from my laptop. here is the new vnc address:
> vnc://dhcp-vl2041-5009.edurom.cornell.edu

(it's just the 5009 part that is new)

password is the same.

let me know if that doesn't solve the problem. thanks!

**Comment by tgerarden** [comment]
just in case you had been using the ip address and not the vnc address, the current ip address is:
> 10.48.20.145

**Comment by tgerarden** [comment]
my laptop and mobile eduroam licenses were all updated on jan 31 so maybe that reset the addresses. this is an annoying aspect of dynamic addresses with wifi but the wired internet in my office is not as reliable as wifi and i've never looked into a fix to get a static address because it doesn't seem to change very often

**Comment by hapham-econ** [comment]
@tgerarden Thank you for the help. I have calculated the weighted elasticities. What do you mean by "assign that to alpha and join it to merged_data"?

**Comment by tgerarden** [comment]
@hapham-econ i was just going based on how the code was written before -- matias had created an object `alpha` to store the demand elasticity. (see line 345 of `estimate_model.R`). that was a constant elasticity, so a scalar / numeric vector of length 1, so we could just use it throughout the code without worrying about its relationship to the data we used it with.

now since it's time-varying we need to make sure we use the right elasticity for a given time period. i was just suggesting that we merge the elasticities into the data frame `merged_data` which already exists to simplify that usage. in hindsight don't worry about adding it to `merged_data` itself. if you create a data frame `alpha` with two columns -- `date` and `alpha` (where alpha is the weighted average elasticity -- you can just merge it into the data in 8) by replacing:

```
cost_est_baseline <- merged_data |> 
  mutate(gross_mc = ifelse(shipments == 0, NA, w * (1 + manuf_share / alpha))) |>
```

with
```
cost_est_baseline <- merged_data |> 
  left_join(alpha) |> 
  mutate(gross_mc = ifelse(shipments == 0, NA, w * (1 + manuf_share / alpha))) |>
```

if you want to use a different naming convention for the dataframe to avoid confusion between alpha the elasticity and alpha the data frame with elasticities, that is fine. but please stick with alpha as the name of the variable within the data frame because it is used in several places throughout the code, not just on line 433. (and if you want to remove the data frame alpha after the join, that's fine, i don't think we'll need it anywhere except that data frame cost_est_baseline since the counterfactuals build off that)

make sense?

**Comment by hapham-econ** [comment]
@tgerarden Here is a histogram of the `gross_ms` estimates:
![image](https://github.com/tgerarden/solar-tariffs-microfounded-demand/assets/90290294/19d479aa-7080-45e3-8ef6-f506c295b856)

and here is a plot for each firm:

![image](https://github.com/tgerarden/solar-tariffs-microfounded-demand/assets/90290294/555e8b0e-ff99-4702-b367-377851fbf7fc)

I haven't figured out how to fix the x-axis yet. Sorry for the late respond, the remote computer thing is not super smooth :(

**Comment by tgerarden** [comment]
@hapham-econ thanks, this is super helpful. can you also post a histogram of the weighted average elasticities, or just a plot of the weighted average elasticity over time, when you get a chance?

sorry to hear the remote computer is not going smoothly, if it continues to be an issue let me know so we aren't held back by tech

**Comment by tgerarden** [comment]
@hapham-econ actually disregard that last ask -- i can just run your code to answer my question in a bit, you can focus on other stuff

**Comment by hapham-econ** [comment]
@tgerarden ok, let me know what to do next!

**Comment by tgerarden** [comment]
@hapham-econ will do. fyi i made a bunch of quicky, sloppy edits to the script you were working on to make some plots to share with coauthors. we'll try to discuss them soon and decide how to move forward on this issue

**Comment by tgerarden** [comment]
@hapham-econ i'm going to meet with my coauthors to discuss next steps thursday. before then, though, would you please look into the estimation error on line 577 of `estimate_model.R`?

it fails for theta = 10. i'm pretty sure this has to be an artifact of the new demand elasticities, but without digging into the data and/or writing down the math to derive the gradient i'm not 100% sure what's driving it and how sensitive estimation would be to using slightly different elasticities (this is not an issue i've run into yet for the larger constant elasticities).

i think i already shared the draft and notes about the estimating equation with you back in january, but let me know if you need anything to guide your investigation into this. thanks!

**Comment by hapham-econ** [comment]
@tgerarden I am having the same trouble connecting with the remote computer as before. Was there any change in the IP address?

**Comment by tgerarden** [comment]
@hapham-econ thanks for the heads up, the computer wasn't connected to wifi because i needed to manually accept a/the new certificate. now it is, but the address changed:

> vnc://dhcp-vl2041-15883.edurom.cornell.edu

> 10.48.63.11

let me know if you run into a problem. if this recurs i will reach out to IT to see if we can take a different approach. thanks!

**Comment by hapham-econ** [comment]
@tgerarden Ahh I see it now. I will continue looking at why the case theta = 5 doesn't converge

**Comment by tgerarden** [comment]
@hapham-econ thanks. hmm. was that error displaying for you when running the code or only after using some post-estimation command to report back errors?

i haven't seen it before -- it's not the error i was getting. maybe it's because of differences in the data we are using (you have access to data through 2018, i have access to data through 2020). here is the error i was getting, which was for `theta=10`:

![Screenshot 2024-02-06 at 4 19 05 PM](https://github.com/tgerarden/solar-tariffs-microfounded-demand/assets/38533138/a9807d7b-5fd3-4e8a-bb3d-52da5d26f5a1)

perhaps we can get around this with alternative starting values, but it's an issue we didn't run into before. but given the difference in data, maybe the "applied" way of approaching this isn't going to be efficient/effective.

based on the error description it seems like this is caused by the estimating equation being non-differentiable at the starting values. could you look into the documentation and see whether that's right?

if it's straightforward to just derive the gradient/jacobian/hessian/whatever is causing the error, let's do that to see whether that helps clarify the source of the problem.

**Comment by hapham-econ** [comment]
@tgerarden Sure, I will look into it now. FYI, I don't have any issue with line 577. When I run line 596 to see the results, this is what I got:

![image](https://github.com/tgerarden/solar-tariffs-microfounded-demand/assets/90290294/a686f6da-8af5-4721-b9c1-41e03b120e10)

I thought the non-convergent issue was with theta = 5.

**Comment by tgerarden** [comment]
@hapham-econ got it, thanks, that's interesting. must just be a difference in the data but i don't have intuition for what the difference is that's driving it. though it is the later periods (which are not in the data you are working with) that are more affected by the switch from aggregate to microfounded demand elasticities. so it must be somehow related to that.

we'll definitely want to make sure all the models converge in the end (which is why we included it in the table) but first help me try to sort out the error above, please!

**Comment by hapham-econ** [comment]
@tgerarden I think it is pretty straightforward to manually derive the Jacobian (I just tried to write it down). The document doesn't specifically mention the error, but based on a few google searches, it seems like a good solution is to change the starting values. Since I don't run into the same error with the 2018 data, the model itself is fine, I think.

**Comment by tgerarden** [comment]
@hapham-econ without knowing the source of the problem, it's hard to know if starting values will solve it. i have tried a few other starting values and that did not resolve the problem. i'd prefer to know the underlying problem rather than manually or automatically search for starting values that may solve the problem (especially if they don't exist for certain values of data...). would you please look into it a bit further to see if you can identify what matrix would cause this error, and then derive it for our estimating equation?

**Comment by tgerarden** [comment]
@hapham-econ okay with a little more experimentation i seem to have found some starting values that work for now, and produce results that seem to make sense at a glance. but i would still prefer to develop and understanding of the problem and would appreciate you looking into it a bit further. if you can't find more useful information after spending another ~1 hour on it we can move on and then revisit in the future as needed

FYI i made the change to `maxfev` that you suggested above. thanks for catching that failure to converge and proposing an effective solution!

**Comment by tgerarden** [comment]
regarding other tasks this week, i'll follow up late thursday or friday this week after speaking with my coauthors. if you'd like to do more RA work in the interim, you could work on [this issue](https://github.com/tgerarden/wto-industrial-policy/issues/2). but it's fine with me if you wait until hearing more instructions on this project since that's a higher priority for me at the moment. i think we'll ask you to move forward with coding up counterfactuals but just want to double-check to make sure the whole team is on board first

**Comment by tgerarden** [comment]
@hapham-econ our meeting was postponed to monday because one of my coauthors is sick.

FYI i learned that the numbers contained in `elasticities.xlsx` are semi-elasticities, not elasticities. i modified the code you had written to convert them to elasticities.

if you are still planning to do RA work on sunday, would you please work on [this issue](https://github.com/tgerarden/wto-industrial-policy/issues/2) while you wait for more instructions on this?

**Comment by hapham-econ** [comment]
@tgerarden Sure, I will work on the other issue!

**Comment by tgerarden** [comment]
@hapham-econ we are still in a holding pattern on fully implementing this issue. we've decided to start by writing up the results we have using the constant elasticity and use the microfounded demand model as a robustness check. so we still want to do it but it is not the top of the priority list right now.

i still need some more info from coauthors to determine the best way to implement it, but we can probably start on it friday or early next week. in the interim i need to focus on other stuff.

one minor thing you could do in the interim is to add one more parameter to the estimating equation to allow some more flexibility in fitting the data:
- add a condition like `include_stockpiling_elasticity = TRUE # FALSE` at the start of `estimate_model.R`
- when false, use current estimating equation
- when true, modify the current estimating equation by replacing the weighted average elasticity `alpha` (which we have estimated and are treating as data) with the term `alpha + kappa` (where `kappa` is a time-invariant parameter we want to estimate. right now the code uses `alpha` in an earlier step so it doesn't show up directly in the estimating equation, but it should be pretty straightforward to combine and write down the alternative estimating equation
- for now include `kappa` in `supply_estimates.tex` when true and post a screenshot of the estimates here on github

thanks!

**Comment by hapham-econ** [comment]
@tgerarden I have a few questions:
- Can I have the new IP address to log into your computer? 
- This `kappa` parameter is just a placeholder for something we want to estimate in the future, right? I will change the part for we define `alpha` in part 6 of the code.
- What is this `supply_estimates.tex` file? Is it a new file that I should create?

**Comment by tgerarden** [comment]
> @tgerarden I have a few questions:
> 
> * Can I have the new IP address to log into your computer?

i'm not sure if/why the IP address keeps changing. it's currently `10.48.63.11`. but would you please try using the vnc address instead? i've been using the address `dhcp-vl2041-15883.eduroam.cornell.edu` since before you said the IP address changed a time or two ago, and it has continued to work without changing... (though it may or may not require the prefix `vnc://`, i'm not sure how it works coming from a pc)

i'm installing cornell-required updates right now and will let you know when they're done

> * This `kappa` parameter is just a placeholder for something we want to estimate in the future, right? I will change the part for we define `alpha` in part 6 of the code.

no, i'm asking you to estimate it right now! please revise the code (really duplicate, since i asked to do it as an option) and then include a `kappa` term as an unknown parameter of the model, estimate it, and report the estimates. you should be able to write the code via dropbox without accessing the computer connection but obviously you'll need the computer for debugging and estimating

> * What is this `supply_estimates.tex` file? Is it a new file that I should create?

it's a file that already exists, it is output by the current script `estimate_model.R` on line 631

**Comment by tgerarden** [comment]
@hapham-econ okay, the computer is ready to go again, thanks for your patience. see above in case you didn't get notified (i forgot to tag). let me know if you have any follow up questions!

**Comment by hapham-econ** [comment]
@tgerarden 
> i'm not sure if/why the IP address keeps changing. it's currently `10.48.63.11`. but would you please try using the vnc address instead? i've been using the address `dhcp-vl2041-15883.eduroam.cornell.edu` since before you said the IP address changed a time or two ago, and it has continued to work without changing... (though it may or may not require the prefix `vnc://`, i'm not sure how it works coming from a pc)
> 
> i'm installing cornell-required updates right now and will let you know when they're done

I tried to use the address above but it doesn't work. I can access the computer by the IP address for now. Last time I tried it was the same IP address but it doesn't work. 

> no, i'm asking you to estimate it right now! please revise the code (really duplicate, since i asked to do it as an option) and then include a `kappa` term as an unknown parameter of the model, estimate it, and report the estimates. you should be able to write the code via dropbox without accessing the computer connection but obviously you'll need the computer for debugging and estimating

I am still having a hard time understanding this `kappa` parameter. Before, I replaced the `alpha` parameter estimated directly in the `estimate_model.R` by the utility + commercial and residential elasticity from 2 other files. Which estimation equation should I change to include this `kappa` parameter? This is also an option inside the option `weighted_elasticity` we had before, right?

**Comment by tgerarden** [comment]
@hapham-econ okay, glad it's working now at least, hopefully the IP address thing will not continue to be a problem.

here is a brief summary of what i'm asking for with equations and references to relevant scripts (though not line numbers):

<img width="881" alt="Screenshot 2024-02-13 at 9 00 55 PM" src="https://github.com/tgerarden/solar-tariffs-microfounded-demand/assets/38533138/ded3b66e-a430-47e9-97ff-bef2b449e57d">

one annoying and potentially confusing thing: the term `alpha` we are using in the code is the same as the term `epsilon^D` in the slides.

does that make sense? or is it still not clear?

and yes, at this point we are only thinking about this as an option inside the option `weighted_elasticity`, thanks for checking! though if it is easy to code it in a `weighted_elasticity`-agnostic way we might as well do that. but if that's not straightforward just do it for `weighted_elasticity==TRUE`.

**Comment by hapham-econ** [comment]
@tgerarden ahhhh it makes sense now! Previously I thought we were trying to estimate another `alpha` to plug into the model. I will start working on this.

**Comment by hapham-econ** [comment]
@tgerarden I spent some time looking at what is the best way to go around this problem. In the end, I create a column in the dataframe (all values are just `TRUE`) which is passed into the `non_linear_model` function. 

To my understanding, it seems like the formula in R doesn't work when I put an independent variable (`kappa` in our case) on the left-hand side. I add a term with `kappa` on the right-hand side instead. 
![image](https://github.com/tgerarden/solar-tariffs-microfounded-demand/assets/90290294/21c8ae4e-3175-47f8-9171-6a64b934245b)
I had some numerical issues when `kappa` was in both the denominator and the numerator, but everything seems good now when I only include `kappa` in the denominator. (it's a minor thing but I was so happy I figured it out)

Here are the results:
![image](https://github.com/tgerarden/solar-tariffs-microfounded-demand/assets/90290294/8fc611be-ba7a-42ba-8af3-fbf0dfeadec2)

I haven't changed the Tex file yet. Let me know what you think and I will adjust accordingly.

**Comment by hapham-econ** [comment]
@tgerarden I am tagging you just in case you missed my last comment. Let me know what you think and what I should do next.

**Comment by tgerarden** [comment]
@hapham-econ thanks. see below.

> @tgerarden I spent some time looking at what is the best way to go around this problem. In the end, I create a column in the dataframe (all values are just `TRUE`) which is passed into the `non_linear_model` function.

why not just have this be a separate object with a value of TRUE/FALSE that is passed as an argument to the relevant function(s)?

> To my understanding, it seems like the formula in R doesn't work when I put an independent variable (`kappa` in our case) on the left-hand side. I add a term with `kappa` on the right-hand side instead. ![image](https://private-user-images.githubusercontent.com/90290294/304882673-21c8ae4e-3175-47f8-9171-6a64b934245b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgzNzAzODAsIm5iZiI6MTcwODM3MDA4MCwicGF0aCI6Ii85MDI5MDI5NC8zMDQ4ODI2NzMtMjFjOGFlNGUtMzE3NS00N2Y4LTkxNzEtNmE2NGI5MzQyNDViLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE5VDE5MTQ0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBmMzFjMjQ1MjljM2UxNTdmOGVlMTc3MDI1NTViNjQwNzRhYmFjMDY0YzBlMDlhN2E0Yjk1OWZhMjU3YTBkN2QmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.wytH2imlyWH-F2ABnUfXccw5Ydgd4A9eFH-WHLDiEt8) I had some numerical issues when `kappa` was in both the denominator and the numerator, but everything seems good now when I only include `kappa` in the denominator. (it's a minor thing but I was so happy I figured it out)

hmm, interesting, i believe stata allows this. but putting that aside, could you write out the steps to that first equality? without taking time to sit down and work it out, i don't follow it.

> Here are the results: ![image](https://private-user-images.githubusercontent.com/90290294/304883204-8fc611be-ba7a-42ba-8af3-fbf0dfeadec2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDgzNzAzODAsIm5iZiI6MTcwODM3MDA4MCwicGF0aCI6Ii85MDI5MDI5NC8zMDQ4ODMyMDQtOGZjNjExYmUtYmE3YS00MmJhLThhZjMtZmJmMGRmZWFkZWMyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMTklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjE5VDE5MTQ0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBkYzEwNTVhYWZhMmZmMjYyYmQ5ZDcyYTYwODBlMjFkMzgxYTg0N2YzNDMxNTg4NWY0M2I3YjMzYjY1YjU0NGImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.oGhzXWqp3fpIQSboEjCUGsr_IAHBex3cCedwfNZu0jY)
> 
> I haven't changed the Tex file yet. Let me know what you think and I will adjust accordingly.

after explaining the derivation above, i'm curious what the estimates look like if we focus on years 2014+ and don't allow for time-varying Ts. i want to know if they are more precise or still super noisy. i added a couple lines to the code to compute this myself but ran into an issue with the starting values again. i also got that when i tried on my machine using the full data. could you debug that and then post an update?

**Comment by hapham-econ** [comment]
@tgerarden 
> why not just have this be a separate object with a value of TRUE/FALSE that is passed as an argument to the relevant function(s)?

I was thinking that `kappa` is an option inside the `weighted_elasticities` option, which is not one of the option in the function but rather changes the input dataframe, so I did the same thing for consistency.

> hmm, interesting, i believe stata allows this. but putting that aside, could you write out the steps to that first equality? without taking time to sit down and work it out, i don't follow it.

After writing it down, I realized I made a mistake. Here is a more detailed version of the derivation. I need to add the second term on the right hand side to the formula we had before.

![image](https://github.com/tgerarden/solar-tariffs-microfounded-demand/assets/90290294/f8185af2-1317-47a2-a693-a726c923b4f2)

> after explaining the derivation above, i'm curious what the estimates look like if we focus on years 2014+ and don't allow for time-varying Ts. i want to know if they are more precise or still super noisy. i added a couple lines to the code to compute this myself (see line 594) but ran into an issue with the starting values again. i also got that when i tried on my machine using the full data. could you debug that and then post an update?

After fixing my code, here are the updated results:

![image](https://github.com/tgerarden/solar-tariffs-microfounded-demand/assets/90290294/fd658b34-f4d8-477f-8cc2-020a6f2f595d)

**Comment by tgerarden** [comment]
@hapham-econ thanks. those estimates make a lot more sense at first glance. i have to focus on course prep and teaching until tomorrow afternoon but then i'll go through the derivation and follow up with you about next steps. sorry for the delay in setting you up with tasks, it's been a busy 1.5 weeks!

**Comment by hapham-econ** [comment]
@tgerarden No worries at all! Have a great rest of the day!

**Comment by tgerarden** [comment]
okay i think we are almost done with this issue for now. we might have to do more work to make everything compatible with the parallel workflow using an aggregate demand model later.

but for now please do a couple things before starting on #2:
- [ ] modify `estimate_model.R` so that we don't estimate and output constant elasticity demand estiamtes when `weighted_elasticity == TRUE`
- [ ] modify `estimate_model.R` so that all the results that are currently output to `supply_estimates` have filenames that distinguish between the "main" version (what we currently have) and an alternative "stockpiling" version when `include_stockpiling_elasticity==TRUE`. you can leave filenames as is for the "main" version. then for the "stockpiling" version just append "stockpiling" to the filename (or we could nest it in another level of directories, but at this point the directory structure needs to be pared back not expanded 😅 
  - then run the estimation code (including outputting the table and plots) both ways for `include_stockpiling_elasticity`
- [ ] let me know when you're done. thanks!

**Comment by tgerarden** [comment]
@hapham-econ FYI i am still making some minor tweaks to the code, i will let you know when i'm done and out so you can start work without worrying about conflicts

**Comment by tgerarden** [comment]
@hapham-econ okay i'm done, will try to stay out of your code now!

**Comment by hapham-econ** [comment]
@tgerarden > * [x]  modify `estimate_model.R` so that we don't estimate and output constant elasticity demand estiamtes when `weighted_elasticity == TRUE`

I included a condition in the part 6 (line 367) to not save the constant elasticities estimate when we use weighted elasticity. Is there anything else you need?

> * [x]  modify `estimate_model.R` so that all the results that are currently output to `supply_estimates` have filenames that distinguish between the "main" version (what we currently have) and an alternative "stockpiling" version when `include_stockpiling_elasticity==TRUE`. you can leave filenames as is for the "main" version. then for the "stockpiling" version just append "stockpiling" to the filename (or we could nest it in another level of directories, but at this point the directory structure needs to be pared back not expanded 😅
>   
>   * then run the estimation code (including outputting the table and plots) both ways for `include_stockpiling_elasticity`

I modified the output and the plot in part 9 and 10 to add the "stockpiling" to the end of the file if we use this option. However, I couldn't run the code because I ran into the non-singular matrix error with `theta = 10`. I tried a couple other starting values for `kappa` but nothing works. Do you run into this problem using your data?

**Comment by tgerarden** [comment]
thanks @hapham-econ 

- the demand estimates are fine. i looked and then decided to rearrange the code a bit so it only estimates the constant elasticity model if it needs to
- right now the code runs with the full data when `include_stockpiling_elasticity==TRUE`
- right now the code fails with the full data when `include_stockpiling_elasticity==FALSE`, producing the following error:

![Screenshot 2024-02-23 at 11 55 53 AM](https://github.com/tgerarden/solar-tariffs-microfounded-demand/assets/38533138/566c9e38-307c-45fe-824d-3f22d211b24f)

**Comment by hapham-econ** [comment]
@tgerarden Do you want me to look into this? Or are you working on it?
> ![Screenshot 2024-02-23 at 11 55 53 AM](https://private-user-images.githubusercontent.com/38533138/307393862-566c9e38-307c-45fe-824d-3f22d211b24f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg3MDc4NzksIm5iZiI6MTcwODcwNzU3OSwicGF0aCI6Ii8zODUzMzEzOC8zMDczOTM4NjItNTY2YzllMzgtMzA3Yy00NWZlLTgyNGQtM2YyMmQyMTFiMjRmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjIzVDE2NTkzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWZjYjA4MGI5YzE4YmE0OGJlZmZhMzFlYzBlYzNiNmJlOGVmNWZlNzIzZTY4NDdmZDliMTI2YmNkMDY0OWY1MzQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.rudUe1KSC8FZsFbnunM_jigDJvzfvS4QsjlHDyds7D8)

**Comment by tgerarden** [comment]
@hapham-econ i would like you to work on it, i am out of this code and need to focus on other things. sorry for not being clear. thanks!

**Comment by hapham-econ** [comment]
@tgerarden Working on it! I was just making sure that we are not both fixing the code at the same time.

**Comment by hapham-econ** [comment]
@tgerarden Fixed. The code should run for the case `include_stockpiling_elasticity==FALSE` now. Let me know if it works on your machine.

**Comment by tgerarden** [comment]
@hapham-econ it runs now, thanks!

**Comment by tgerarden** [comment]
i think this issue is basically complete but i am going to leave it open for now so we remember to come back and document anything we need from this issue in the paper and/or readme files
